### ISSUE TITLE
weird reportUnboundVariable behaviour

### ISSUE THREAD

**Issue by klement** [microsoft/pyright#3041]
So this construct caught my eye:

```python
  def x():
      y = True
      if y:
          a = 1
      if y:
>>        print(a)
                 "a" is possibly unbound
                 [Pyright: reportUnboundVariable]
                 ────────────────────────────────────────────────────────────────────
                 https://github.com/microsoft/pyright/blob/main/docs/configuration.md
```

I don't think there is a way how a could be unbound ...

curiously, it also complains for x

```python
  def x():
      y = True
      if y:
          x = 1
      if y:
>>        print(x)
                 "x" is possibly unbound
                 [Pyright: reportUnboundVariable]
                 ────────────────────────────────────────────────────────────────────
                 https://github.com/microsoft/pyright/blob/main/docs/configuration.md
```

but in this case x is surely at least the function itself, yes?

changing the variable name makes it see that x can be the function itself:

```
  def x():
      y = True
      if y:
>>        a = 1
      if y:
          print(x)
```

there is now warning "a" is not accessed, but it no longer complains about x being possibly unbound

**Comment by erictraut** [comment]
> I don't think there is a way how a could be unbound

Pyright tracks the types of individual variables. It doesn't track types of one variable (such as `a`) conditioned on the type of another variable (such as `y`). I'm not aware of any static analyzer or compiler in any language that does that. So you're expecting too much of a static type checker here. You can work around this by always initializing `a` along all code paths.

```python
def x():
    a = None
    y = True

    if y:
        a = 1

    if y:
        print(a)
```

Alternatively, if `y` is really always True, then there's no need to include the conditional checks in the first place, but I presume you did that simply to reduce the code to a minimal sample.

> but in this case x is surely at least the function itself

No, the `x` within the function scope is different from the `x` outside the function.

```python
def x():
    print(x) # Runtime exception
    x = 1

x()
```

So, everything here is expected behavior. I'm going to close the issue.

**Comment by klement** [comment]
Thanks for looking into this. If it's as designed then there is no contest.

I'm pretty sure that coverity (for C) does this, so that's why I thought pyright might too.

Yes, it's a minimal example and adding those initialisations for the sake of static checker is not always desirable.

Regarding your last statement about scopes, it depends whether x is also used locally or not. So in my example, x is the functions itself. In yours, there will be a runtime exception...

```
Klements-MacBook-Air:~ ksekera$ python3
Python 3.9.9 (main, Nov 21 2021, 03:16:13)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.


>>> def x():
...     print("x is:")
...     print(x)
...
>>> x()
x is:
<function x at 0x1005ab4c0>


>>> def x():
...     print(x)
...     x = 5
...     print(x)
...
>>> x()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in x
UnboundLocalError: local variable 'x' referenced before assignment
```

**Comment by JelleZijlstra** [comment]
> Regarding your last statement about scopes, it depends whether x is also used locally or not.

No, it depends on whether there is any assignment to `x` in the inner scope. In the second example in your original post, `x` within the function cannot refer to the function, because there is a local variable named `x`.

**Comment by klement** [comment]
My bad, I wrote used but meant assigned 😊
